build-support/bin/download_binary.sh

```diff
@@ -25,16 +30,43 @@ readonly util_name="$2"
 readonly version="$3"
 readonly filename="${4:-${util_name}}"
 
-os=$("${REPO_ROOT}/build-support/bin/get_os.sh")
-readonly path="bin/${util_name}/${os}/${version}/${filename}"
+function make_download_path_for_os {
+  local os_string="$1"
+  echo "bin/${util_name}/${os_string}/${version}/${filename}"
+}
+
+function curl_output_file_with_fail {
+  local download_url="$1"
+  local output_path="$2"
+  echo >&2 "Downloading ${download_url} ..."
+  curl --fail "$download_url" > "$output_path"
+}
+
+path="$(make_download_path_for_os "$("${REPO_ROOT}/build-support/bin/get_os.sh")")"
 readonly cache_path="${CACHE_ROOT}/${path}"
 
+maybe_alternate_os="$([[ "$(uname)" == 'Darwin' ]] \
+    && echo "mac/${MOST_RECENT_SUPPORTED_MACOS_VERSION}")"
+
 if [[ ! -f "${cache_path}" ]]; then
   mkdir -p "$(dirname "${cache_path}")"
 
   readonly binary_url="https://${host}/${path}"
-  echo >&2 "Downloading ${binary_url} ..."
-  curl --fail "${binary_url}" > "${cache_path}".tmp
+  readonly output_tmp_file="${cache_path}.tmp"
+  readonly curl_err_tmp_file="${cache_path}.curl.err.tmp"
+
+  curl_output_file_with_fail "${binary_url}" "$output_tmp_file" \
+                             2> >(tee >&2 "$curl_err_tmp_file") \
+    || if grep -q -F 'The requested URL returned error: 404' "$curl_err_tmp_file" \
```

Comment:
Do `-q` and `-F` have long flags?

Reply:
They do (I think `--quiet`/`--silent` and `--fixed-strings`?). They should absolutely have been used here.